tools/python/util/reduced_build_config_parser.py

```diff
@@ -0,0 +1,123 @@
+
+    If for example the types of inputs 0 and 1 were important, the entry may look like this (e.g. ai.onnx:Gather):
+        `{"inputs": {"0": ["float", "int32_t"], "1": ["int32_t"]}}`
+
+    Finally some operators do non-standard things and store their type information under a 'custom' key.
+    ai.onnx.OneHot is an example of this, where 3 type names from the inputs are combined into a string.
+        `{"custom": ["float_int64_t_int64_t", "int64_t_string_int64_t"]}`
+
+    :param config_file: Configuration file to parse
+    :param enable_type_reduction: Set to True to use the type information in the config.
+                                  If False the type information will be ignored.
+                                  If the flatbuffers module is unavailable any type information will be ignored
+                                  as the usage of the type information is via OperatorTypeUsageManager which has a
+                                  dependency on the ORT flatbuffers python schema.
+    :return: required_ops, op_type_usage_manager:
+             Dictionary of domain:opset:[ops] for required operators
+             OperatorTypeUsageManager manager with operator specific type usage information if available. None if
+             type reduction was disabled, or the flatbuffers module is not available.
+    '''
+
+    if not os.path.isfile(config_file):
+        raise ValueError('Configuration file {} does not exist'.format(config_file))
+
+    required_ops = {}
+    op_type_usage_manager = OperatorTypeUsageManager() if enable_type_reduction and have_flatbuffers else None
+
+    with open(config_file, 'r') as config:
+        for line in [orig_line.strip() for orig_line in config.readlines()]:
+            if not line or line.startswith("#"):  # skip empty lines and comments
+                continue
+
+            domain, opset_str, operators_str = [segment.strip() for segment in line.split(';')]
+            opset = int(opset_str)
+
+            # any type reduction information is serialized json that starts/ends with { and }.
+            # type info is optional for each operator.
+            if '{' in operators_str:
+                # parse the entries in the json dictionary with type info
+                operators = set()
+                cur = 0
+                end = len(operators_str)
+                while cur < end:
+                    next_comma = operators_str.find(',', cur)
+                    next_open_brace = operators_str.find('{', cur)
+
+                    if next_comma == -1:
+                        next_comma = end
+
+                    # the json string starts with '{', so if that is found (next_open_brace != -1)
+                    # before the next comma (which would be the start of the next operator if there is no type info
+                    # for the current operator), we have type info to parse
+                    if 0 < next_open_brace < next_comma:
```

Comment:
nit, can we add an example string here to better explain the parsing?

Reply:
There are a few examples in the function description. Is that not enough?

Reply:
This is probably for other people try to understand the code, but yes we can look at the example string above, just though with something next to it will be covenient :smile: 

---
In reply to: [565065055](https://github.com/microsoft/onnxruntime/pull/6428#discussion_r565065055) [](ancestors = 565065055)